fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/synchronizedsession/SynchronizedSessionRunner.java

```diff
@@ -0,0 +1,218 @@
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb.synchronizedsession;
+
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.provider.common.StoreTimer;
+import com.apple.foundationdb.record.provider.foundationdb.FDBDatabase;
+import com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseRunner;
+import com.apple.foundationdb.record.provider.foundationdb.FDBDatabaseRunnerInterface;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordContext;
+import com.apple.foundationdb.record.provider.foundationdb.FDBStoreTimer;
+import com.apple.foundationdb.subspace.Subspace;
+import org.apache.commons.lang3.tuple.Pair;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * It delegates all methods to {@link FDBDatabaseRunner}. For {@code run} and {@code runAsync} methods that take the
+ * work {@code retriable}, it injects necessary operations to keep the session synchronized by wrapping the work} with
+ * {@link SynchronizedSession#workInSession(Function)} or {@link SynchronizedSession#workInSessionAsync(Function)}.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class SynchronizedSessionRunner implements FDBDatabaseRunnerInterface {
+
+    private FDBDatabaseRunner underlying;
+    private SynchronizedSession session;
+
+    // Start a new session. A synchronized session keeps its lock by updating the timestamp in each of its working
+    // transactions, so do NOT try to get the synchronized runner until it is going to be actively used.
+    public static SynchronizedSessionRunner startSession(@Nonnull Subspace lockSubspace,
+                                                         long leaseLengthMill,
+                                                         @Nonnull FDBDatabaseRunner runner) {
+        final UUID newSessionId = UUID.randomUUID();
+        SynchronizedSession session = new SynchronizedSession(lockSubspace, newSessionId, leaseLengthMill);
+        try (FDBRecordContext context = runner.openContext()) {
```

Comment:
I can't recall why this is being done in another transaction.  Does it really need to be?  There should probably be a  java doc describing the rationale for this decision.